### PR TITLE
Add a lock to ProxyTypeEmitter

### PR DESCRIPTION
**Pull request by jbagga**
Mitigate race conditions in https://github.com/aspnet/Hosting/issues/1312

### CODE REVIEW

File: test/Microsoft.Extensions.DiagnosticAdapter.Test/Internal/ProxyTypeEmitterTest.cs [hunk 1]
```diff
@@ -623,6 +624,22 @@ public void Adapt_NestedList()
...
+                0,
+                100,
+                (j) =>
+                {
+                    var testObject = new Person();
+                    ProxyTypeEmitter.GetProxyType(new ProxyTypeCache(), typeof(IPerson), testObject.GetType());
+                });
+            }
+        }
+
```
Comment (jbagga) [inline test/Microsoft.Extensions.DiagnosticAdapter.Test/Internal/ProxyTypeEmitterTest.cs]: This test fails consistently without the lock statements in `ProxyTypeEmitter`

Comment (jbagga) [reply test/Microsoft.Extensions.DiagnosticAdapter.Test/Internal/ProxyTypeEmitterTest.cs]: This is the exception thrown: 
```
Message: System.AggregateException : One or more errors occurred. (Duplicate type name within an assembly.)
---- System.ArgumentException : Duplicate type name within an assembly.
```

Comment (jbagga) [reply test/Microsoft.Extensions.DiagnosticAdapter.Test/Internal/ProxyTypeEmitterTest.cs]: Stack Trace:
```
Failed   Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitterTest.GetProxyType_LocksPreventDuplicateAssemblyNamesArgumentException_ForConcurrentThreads
Error Message:
 System.AggregateException : One or more errors occurred.
---- System.ArgumentException : Duplicate type name within an assembly.
Stack Trace:
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, Action`1 body)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitterTest.GetProxyType_LocksPreventDuplicateAssemblyNamesArgumentException_ForConcurrentThreads() in C:\git\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Internal\ProxyTypeEmitterTest.cs:line 632
----- Inner Stack Trace -----
   at System.Reflection.Emit.ModuleBuilder.CheckTypeNameConflict(String strTypeName, Type enclosingType)
   at System.Reflection.Emit.AssemblyBuilderData.CheckTypeNameConflict(String strTypeName, TypeBuilder enclosingType)
   at System.Reflection.Emit.TypeBuilder.Init(String fullname, TypeAttributes attr, Type parent, Type[] interfaces, ModuleBuilder module, PackingSize iPackingSize, Int32 iTypeSize, TypeBuilder enclosingType)
   at System.Reflection.Emit.ModuleBuilder.DefineType(String name, TypeAttributes attr, Type parent, Type[] interfaces)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyAssembly.DefineType(String name, TypeAttributes attributes, Type baseType, Type[] interfaces)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitter.GenerateProxyTypeFromProperties(Type sourceType, Type targetType, VerificationResult verificationResult)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitter.VerifyProxySupport(ProxyBuilderContext context, Tuple`2 key)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitter.VerifyProxySupport(ProxyBuilderContext context, Tuple`2 key)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitter.GetProxyType(ProxyTypeCache cache, Type targetType, Type sourceType)
   at Microsoft.Extensions.DiagnosticAdapter.Internal.ProxyTypeEmitterTest.<>c.<GetProxyType_LocksPreventDuplicateAssemblyNamesArgumentException_ForConcurrentThreads>b__31_0(Int32 j) in C:\git\EventNotification\test\Microsoft.Extensions.DiagnosticAdapter.Test\Internal\ProxyTypeEmitterTest.cs:line 638
   at System.Threading.Tasks.Parallel.<>c__DisplayClass17_0`1.<ForWorker>b__1()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass176_0.<ExecuteSelfReplicating>b__0(Object )

```

**Comment by Eilon** [comment]
Is this supposed to go into the 2.0.x patch? Is there a bug tracking that?

**Comment by jbagga** [comment]
No I am not sure if we want to patch it. I sent this because right now we target netcoreapp2.0 and net461 only (see [here](https://github.com/aspnet/EventNotification/pull/84)) so it might be useful to patch. Happy to close this and only keep the dev PR if needed

**Comment by jbagga** [comment]
Closing this PR.

**Pull request closed without merging**